accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,260 @@
+  code to create an assembler library ("MacroAssembler")
+* Prioritize compilation performance over runtime performance
+
+### High-level overview
+
+```mermaid
+  graph TD;
+  A(wasmparser)-->B(cranelift-wasm);
+  A-->C(winch);
+  C-->D(MacroAssembler);
+  D-->X(cranelift-asm);
+  X-->E(MachInst);
+  X-->F(MachBuffer);
+  B-->G(cranelift);
+  G-->X;
+```
+
+### MacroAssembler and Borrowing from Cranelift
+
+We plan to factor out the lower layers of Cranelift that produce and operate on
+machine code in order to reuse them as a generic assembler library
+(“MacroAssembler”).
+
+The two key abstractions that will be useful to reuse are the `MachInst`
+(“machine instruction”) trait and its implementations for each architecture; and
+the `MachBuffer`, which is a machine-code emission buffer with some knowledge of
+branches and ability to do peephole optimizations on them. The former lets us
+reuse all the logic to encode instructions for an ISA; the latter lets us emit
+code with “labels” and references to labels, and have the fixups done for us.
+
+The `MachInst` trait and its implementations, and the `MachBuffer`, can be
+mostly factored out into a separate crate `cranelift_asm`. This will require
+some care with respect to layering: in particular, definitions of
+machine-instruction types are currently done in the ISLE backends for each ISA
+within Cranelift. We can continue to use ISLE for these, but they will need to
+be moved to the separate crate.
+
+
+As a result of this initial layering, one will be able to build a `MachInst` as
+a Rust data structure and emit it manually, for example:
+
+```rust
+let add = cranelift_asm::x64::AluRmiR { op: AluRmiR::Add, … }; let mut
+buf = MachBuffer::new(); add.emit(&mut buf, …);
+```
+
+However this is still quite cumbersome. As a next step, we will develop an API
+over this that provides for procedural generation of instructions: i.e., one
+method call for each instruction. Something like:
+
+```rust
+let mut masm = cranelift_asm::x64::MacroAssembler::new(); 
+masm.add(rd, rm);
+masm.store(rd, MemArg::base_offset(ra, 64));
+``` 
+This would allow for
+fairly natural single-pass code emission. In essence, this is an implementation
+of the [MacroAssembler
+idea](https://searchfox.org/mozilla-central/rev/fa71140041c5401b80a11f099cc0cd0653295e2c/js/src/jit/MacroAssembler.h)
+from SpiderMonkey. Each architecture will have an implementation of the
+MacroAssembler API; perhaps there can be a trait that abstracts commonalities,
+but enough will be different (e.g., instruction set quirks beyond the usual
+“add/sub/and/or/not” suspects, x64 two-operand form vs aarch64 three-operand
+form, and more) that we expect there to be different `MacroAssembler` types for
+each ISA. This in turn implies different lowering code that invokes the
+`MacroAssembler` per ISA in the baseline compiler. The lowering code can perhaps
+share many helpers that are monomorphized on the “common ISA core” trait.
```

Comment:
I'd like to echo @alexcrichton's suggestion about building out (at least a realistic kernel of) support for a second architecture during the initial implementation. Because Winch is, by design, skipping an architecture-agnostic mid-end and IR, it has higher risk of baking in architecture-specific assumptions into the core design. Building support for a second architecture should mitigate and nullify this risk.